### PR TITLE
Update our CI pipelines to include Node 18

### PR DESCRIPTION
**Pull request by octogonz**
## Summary

Node 18 has been LTS since 2022-10-25, but our CI pipelines have not been building it because of incompatibility with Webpack 4.

## Details

- `heft-webpack4-plugin` now prints a warning if Node 18 is used without the `NODE_OPTIONS=--openssl-legacy-provider` workaround
- Enable this workaround in our CI pipelines
- Update `NodeJsCompatibility.ts` to specify 18 as the new LTS
- Upgrade rush.json to use PNPM 7.33.5
- I noticed that `autoinstaller.lock` sometimes appears in my Git diff view, so I've added `*.lock` to .gitignore and synced up the `rush init` template

## How it was tested

- [x] Confirm that CI build passes for this PR

@TheLarkInn @iclanton

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -50,11 +54,18 @@ jobs:
           # See https://github.com/microsoft/rushstack/issues/2981
           BROWSERSLIST_IGNORE_OLD_DATA: 1
 
+          # Workaround for Node 18 incompatibility with Webpack 4
+          NODE_OPTIONS: --openssl-legacy-provider
```
Comment (iclanton) [inline .github/workflows/ci.yml:58]: We should only set this if it's specifically necessary.

File: heft-plugins/heft-webpack4-plugin/src/Webpack4Plugin.ts [hunk 2]
```diff
@@ -184,11 +184,31 @@ export default class Webpack4Plugin implements IHeftTaskPlugin<IWebpackPluginOpt
     return this._webpackCompiler;
   }
 
+  private _warnAboutNodeJsIncompatibility(taskSession: IHeftTaskSession): void {
+    const versionMatch: RegExpExecArray | null = /^([0-9]+)\./.exec(process.versions.node); // parse the SemVer MAJOR part
+    if (versionMatch) {
+      const nodejsMajorVersion: number = parseInt(versionMatch[1]);
+      if (nodejsMajorVersion >= 16) {
+        if (!/(^|[^a-z\-])--openssl-legacy-provider($|[^a-z\-])/.test(process.env.NODE_OPTIONS ?? '')) {
```
Comment (iclanton) [inline heft-plugins/heft-webpack4-plugin/src/Webpack4Plugin.ts:192]: Leave a comment describing what this matches.

**Review by iclanton — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (octogonz) [reply .github/workflows/ci.yml]: Fixed

File: heft-plugins/heft-webpack4-plugin/src/Webpack4Plugin.ts (hunk 2 above)
Comment (octogonz) [reply heft-plugins/heft-webpack4-plugin/src/Webpack4Plugin.ts:198]: Fixed

**Pull request merged**